lsp-mode.el

```diff
@@ -4779,6 +4779,23 @@ When language is nil render as markup if `markdown-mode' is loaded."
                               (or title command))
                             nil t))))
 
+(defun lsp--workspace-server-id (workspace)
+  "Return the server ID of WORKSPACE."
+  (-> workspace lsp--workspace-client lsp--client-server-id symbol-name))
+
+(defun lsp--handle-rendered-for-echo-area (contents)
+  "Return a single line from RENDERED, appropriate for display in the echo area."
+  (pcase (lsp-workspaces)
+    (`(,workspace)
+     (if (string-equal "clangd" (lsp--workspace-server-id workspace))
+         (progn
+           (require 'lsp-clients)
```

Comment:
Can we restructure to remove that? I mean, one might not want to include lsp-clients (we have settings to avoid that) and in the end, the require will be forced? Also, the require of files that have lsp mode client registrations is slow due to lsp-register-custom-settings calls.

Reply:
I think we can use `cl-defgeneric/cl-defmethod` here.
For example
``` elisp
(cl-defgeneric lsp-clients-extract-signature-on-hover (server-id contents)
  "Extract a representative line from CONTENTS, to show in the echo area."
  (car (s-lines (lsp--render-element contents))))
```
And put this in `lsp-clients.el`
``` elisp
(cl-defmethod lsp-clients-extract-signature-on-hover ((_server-id (eql clangd)) contents)
  ""
  ;; ...
  )
```
User will have to require `lsp-clients` to use `lsp-clangd` anyway, so the correct specializer will be loaded.